apiserver/facades/client/cloud/cloud.go

```diff
@@ -379,9 +414,125 @@ func (api *CloudAPI) Credential(args params.Entities) (params.CloudCredentialRes
 
 // AddCloud adds a new cloud, different from the one managed by the controller.
 func (api *CloudAPIV2) AddCloud(cloudArgs params.AddCloudArgs) error {
+	// TODO (anastasiamac 2018-02-16) This is needed as facade v2 has CAAS-specific calls
+	// but we need to bump facade version and make v3 available.
+	if !featureflag.Enabled(feature.CAAS) {
+		return errors.NotSupportedf("AddCloud call")
+	}
 	err := api.backend.AddCloud(common.CloudFromParams(cloudArgs.Name, cloudArgs.Cloud))
 	if err != nil {
 		return err
 	}
 	return nil
 }
+
+// CredentialContents returns the specified cloud credentials,
+// including the secrets if requested.
+// If no specific credential name/cloud was passed in, all credentials for this user
+// are returned.
+// Only credential owner can see its contents as well as what models use it.
+// Controller admin has no special superpowers here and is treated the same as all other users.
+func (api *CloudAPIV3) CredentialContents(args params.CredentialContentRequestParam) (params.CredentialContentResults, error) {
+	// Helper to look up and cache credential schemas for clouds.
+	schemaCache := make(map[string]map[cloud.AuthType]cloud.CredentialSchema)
+	credentialSchemas := func(cloudName string) (map[cloud.AuthType]cloud.CredentialSchema, error) {
+		if s, ok := schemaCache[cloudName]; ok {
+			return s, nil
+		}
+		cloud, err := api.backend.Cloud(cloudName)
+		if err != nil {
+			return nil, err
+		}
+		provider, err := environs.Provider(cloud.Type)
+		if err != nil {
+			return nil, err
+		}
+		schema := provider.CredentialSchemas()
+		schemaCache[cloudName] = schema
+		return schema, nil
+	}
+
+	// Helper to parse state.Credential into an expected result item.
+	stateIntoParam := func(credential state.Credential, includeSecrets bool) params.CredentialContentResult {
```

Comment:
this is a big method - i'm not sure in-lining it buys anything - better i think to put it on the api struct

Reply:
It uses cloudschemas that are built dynamically. This is very similar to how you've build CloudCredential method above... I cannot think of a way to break it down.. It only appears long because of a helper functions that are declared within it.